src/betterproto/__init__.py

```diff
@@ -487,6 +490,51 @@ def decode_varint(buffer: bytes, pos: int) -> Tuple[int, int]:
             raise ValueError("Too many bytes when decoding varint.")
 
 
+def read_varint(stream: BinaryIO) -> Optional[int]:
+    """
+    Read a single varint value from a stream. Returns the value or :code:`None` if the
+    stream is already at EOF.
+    """
+    if "b" not in stream.mode:
+        raise TypeError("`stream` must be in binary mode but isn't")
+    elif not stream.readable():
+        raise TypeError("`stream` must be readable but isn't")
+
+    value = 0
+
+    for shift in count():
+        # Read the next byte and check that it is valid
+        byte = stream.read(1)
+        if byte == b"":
+            # EOF - only valid if there are no further expected bytes
+            if shift == 0:
+                # No bytes read yet, so none expected - valid
+                return None
+            else:
+                # Was expecting to read another byte if this was reached - invalid
+                raise ValueError("Stream ended part way through a varint")
+
+        # Add current byte's stored value to the overall value
+        byte_int = int.from_bytes(byte, byteorder="little")
+        value += (byte_int & 0b01111111) << (7 * shift)
+
+        # Stop reading if this is the last byte
+        if byte_int >> 7 != 1:
+            break
+
+    return value
```

Comment:
If you want to be really fancy here:

```suggestion
    def _ints() -> Iterator[int]:
        byte = stream.read(1)
        if not byte:
            raise EOFError("No more bytes for decoding varint in stream")
        yield int.from_bytes(byte, byteorder="little")

    return accumulate(
        (i & 0b01111111) << s
        for i, s in zip(takewhile(i & 0b10000000,_ints(stream)), count(step=7))
    )
```

Need `accumulate`/`takewhile` from `itertools`.

Reply:
Sorry, that code is garbage 😆 

```py
    def _ints() -> Iterator[int]:
        byte = stream.read(1)
        if not byte:
            raise EOFError("No more bytes for decoding varint in stream")
        yield int.from_bytes(byte, byteorder="little")

    def _takeuntil(predicate: Callable[[T], bool], i: Iterable[T]) -> Iterator[T]:
        for v in i:
            yield v
            if predicate(v):
                break

    return accumulate(
        (i & 0b01111111) << s
        for i, s in zip(_takeafter(v: v & 0b10000000,_ints(stream)), count(step=7))
    )
```

That should work but by that point it is about as big as the original `for` loop! Sorry, bad suggestion!